components/file_system/src/thread_io.rs

```diff
@@ -0,0 +1,269 @@
+        Arc::new(ThreadIOSentinel::current_default())
+    ).clone();
+
+    static THREAD_IO_PRIVATE: RefCell<ThreadIOPrivate> = RefCell::new(ThreadIOPrivate::default())
+}
+
+struct ThreadIOSentinel {
+    id: ThreadID,
+    bytes: [AtomicIOBytes; IOType::COUNT],
+}
+
+struct ThreadID {
+    pid: Pid,
+    tid: Pid,
+}
+
+struct ThreadIOPrivate {
+    io_type: IOType,
+    last_bytes: IOBytes,
+}
+
+struct AtomicIOBytes {
+    read: AtomicU64,
+    write: AtomicU64,
+}
+
+pub fn fetch_all_thread_io_bytes(io_type: IOType) -> IOBytes {
+    THREAD_IO_SENTINEL_VEC.iter().for_each(|sentinel| {
+        fetch_sentinel(sentinel, None);
+    });
+    THREAD_IO_TOTAL[io_type as usize].load(Ordering::Relaxed)
+}
+
+fn fetch_sentinel(sentinel: &ThreadIOSentinel, new_io_type: Option<IOType>) {
+    let io_bytes = fetch_exact_thread_io_bytes(&sentinel.id);
+
+    THREAD_IO_PRIVATE.with(|private| {
+        let io_type = private.borrow().io_type;
+        let last_io_bytes = private.borrow().last_bytes;
+
+        sentinel.bytes[io_type as usize].fetch_add(io_bytes - last_io_bytes, Ordering::Relaxed);
+        THREAD_IO_TOTAL[io_type as usize].fetch_add(io_bytes - last_io_bytes, Ordering::Relaxed);
+        private.borrow_mut().last_bytes = io_bytes;
+        if let Some(new_io_type) = new_io_type {
+            private.borrow_mut().io_type = new_io_type;
+        }
+    })
+}
+
+pub fn set_io_type(new_io_type: IOType) {
+    THREAD_IO_SENTINEL.with(|sentinel| {
+        fetch_sentinel(sentinel, Some(new_io_type));
+    })
+}
+
+pub fn get_io_type() -> IOType {
+    THREAD_IO_PRIVATE.with(|private| private.borrow().io_type)
+}
+
+pub(crate) fn fetch_buffered_thread_io_bytes(io_type: IOType) -> IOBytes {
+    THREAD_IO_SENTINEL.with(|sentinel| sentinel.bytes[io_type as usize].load(Ordering::Relaxed))
+}
+
+pub(crate) fn fetch_thread_io_bytes(_io_type: IOType) -> IOBytes {
+    fetch_exact_thread_io_bytes(&ThreadID::current())
+}
+
+fn fetch_exact_thread_io_bytes(id: &ThreadID) -> IOBytes {
+    let io_file_path = PathBuf::from("/proc")
+        .join(format!("{}", id.pid))
+        .join("task")
+        .join(format!("{}", id.tid))
+        .join("io");
+
+    if let Ok(io_file) = File::open(io_file_path) {
+        return IOBytes::from_io_file(io_file);
+    }
+
+    IOBytes::default()
+}
+
+impl ThreadIOSentinel {
+    fn current_default() -> ThreadIOSentinel {
+        ThreadIOSentinel {
+            id: ThreadID::current(),
+            bytes: Default::default(),
+        }
+    }
+}
+
+impl ThreadID {
+    fn current() -> ThreadID {
+        ThreadID {
+            pid: getpid(),
+            tid: gettid(),
+        }
+    }
+}
+
+impl Default for ThreadIOPrivate {
+    fn default() -> ThreadIOPrivate {
+        ThreadIOPrivate {
+            io_type: IOType::Other,
+            last_bytes: IOBytes::default(),
+        }
+    }
+}
+
+impl IOBytes {
+    fn from_io_file<R: std::io::Read>(r: R) -> IOBytes {
```

Comment:
Could you remove this function and inline the code inside `fetch_exact_thread_io_bytes`? I personally think the code is strongly dependent on the file path, therefore shouldn't be split from `fetch_exact_thread_io_bytes`.